### PR TITLE
Improve prompt support in JJ

### PR DESCRIPTION
**Pull request by torquestomp**
Adds `ui.prompt...()` functions and uses them to fix issue #2126

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/next.rs
Comment (PhilipMetzger) [inline cli/src/commands/next.rs:99]: Not blocking, but it could be valuable to move this function to something like `movement_utils.rs` as both `jj edit @+/@-` (the old way of `next/prev`) also runs into this issue.

**Review by PhilipMetzger — APPROVED** [review]
LGTM for `next` & `prev`. I defer to martin for the new test utilities/mini framework.

File: cli/src/commands/next.rs
Comment (torquestomp) [reply cli/src/commands/next.rs:99]: I thought of that too (moving this to a helper) but without a coded 3rd use case I'm not sure what it should look like / what that helper file should be named.  Edit currently calls `resolve_single_rev()` so would need more (likely controversial) work to make use of this.

Comment (PhilipMetzger) [reply cli/src/commands/next.rs:99]: For me it's fine to defer this to later so if you also want to tackle the `jj edit` problem, we can decide then.  

> I thought of that too (moving this to a helper) but without a coded 3rd use case I'm not sure what it should look like / what that helper file should be named.

That's fair.

File: cli/src/ui.rs
Comment (yuja) [inline cli/src/ui.rs:369]: better to terminate on EOF (empty string without `\n`) by mapping to io::Error?

File: cli/src/ui.rs
Comment (yuja) [inline cli/src/ui.rs:379]: nit: `self.warning()` could be used?

File: cli/src/ui.rs
Comment (yuja) [inline cli/src/ui.rs:397]: nit: `starts_with(['y', 'Y'])` will also work

File: cli/src/commands/next.rs
Comment (yuja) [inline cli/src/commands/next.rs:77]: perhaps, `write_commit_summary()` can be used here

File: cli/src/commands/next.rs
Comment (yuja) [inline cli/src/commands/next.rs:87]: nit: `s.as_ref()` or `s.as_str()` also works. Alternatively, `ui.prompt_choice()` could be generalized to accept `&[impl AsRef<str>]` or `Iterator<...>`

**Review by yuja — APPROVED** [review]

File: cli/src/ui.rs
Comment (joyously) [reply cli/src/ui.rs:397]: Why is it not translated?

File: cli/src/ui.rs
Comment (yuja) [reply cli/src/ui.rs:369]: Perhaps, EOF should be handled by `prompt()`, and the trailing `"\n"` should be removed there. I have no idea how libgit2 callback handles "{username}\n".

**Comment by ilyagr** [comment]
Idle thought (for future PRs maybe): I was thinking of trying something like https://github.com/console-rs/dialoguer for this, which might allow a nicer UI for a selection list. A simpler version is almost certainly better at first, though, and who knows how polished that library actually is.

File: cli/src/commands/next.rs
Comment (torquestomp) [reply cli/src/commands/next.rs]: I went with `&[impl AsRef<str>]`; I couldn't figure out a way to use `Iterator` that I liked.  `impl IntoIterator<Item = impl AsRef<str>>` only works well if we use it once, it doesn't work with the `loop {}`.

File: cli/src/commands/next.rs
Comment (torquestomp) [reply cli/src/commands/next.rs]: Done.

File: cli/src/ui.rs
Comment (torquestomp) [reply cli/src/ui.rs:381]: Done, added a new commit for this at the base.

File: cli/src/ui.rs
Comment (martinvonz) [inline cli/src/ui.rs:360]: I think this can be simplified by using [strip_suffix](https://doc.rust-lang.org/std/primitive.str.html#method.strip_suffix)

File: cli/src/ui.rs
Comment (martinvonz) [inline cli/src/ui.rs:359]: Maybe this should be a `user_error()` instead?

File: CHANGELOG.md
Comment (martinvonz) [inline CHANGELOG.md:1]: Please drop the merge commit before merging

File: cli/src/ui.rs
Comment (torquestomp) [reply cli/src/ui.rs:357]: This method returns `io::Result<>`, so we'd need to change that?  I feel like that's out of scope; 'unexpectedEof' is very accurate here.

File: cli/src/ui.rs
Comment (torquestomp) [reply cli/src/ui.rs]: Done.

File: CHANGELOG.md
Comment (torquestomp) [reply CHANGELOG.md:1]: Done.

**Pull request merged**
